File: src/main/java/javabot/web/views/PagedView.java
package javabot.web.views;

import com.google.inject.Injector;
import javabot.model.Factoid;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

public abstract class PagedView<V> extends MainView {
    public static final int ITEMS_PER_PAGE = 50;
    private final Integer itemsPerPage = ITEMS_PER_PAGE;
    private int page;
    private Long itemCount;

    public PagedView(final Injector injector, final HttpServletRequest request, final int currentPage) {
        super(injector, request);
        this.page = currentPage;
    }

    @Override
    public final String getChildView() {
        return "paged.ftl";
    }

    public abstract String getPagedView();

    public final Long getItemCount() {
        if (itemCount == null) {
            itemCount = countItems();
        }
        return itemCount;
    }

    public abstract Long countItems();

    public V getFilter() {
        return null;
    }

    public int getPage() {
        if (page < 1) {
            page = 1;
        } else if (page > getPageCount()) {
            page = getPageCount();
        }
        return page;
    }

    public int getIndex() {
        int index = (getPage() - 1) * getItemsPerPage();
        if (getItemCount() == 0) {
            return -1;
        } else if (index > getItemCount()) {
            return (getPageCount() - 1) * getItemsPerPage();
        } else {
            return index;
        }
    }

    public int getPageCount() {
        return Double.valueOf(Math.ceil(1.0 * getItemCount() / getItemsPerPage())).intValue();
    }

    public int getItemsPerPage() {
        return itemsPerPage;
    }

    public String getNextPage() {
        final int nextPage = getPage() + 1;
        return nextPage <= getPageCount() ? getPageUrl() + "?page=" + nextPage : null;
    }

    protected abstract String getPageUrl();

    public String getPreviousPage() {
        final int previousPage = getPage() == 1 ? -1 : getPage() - 1;
        return previousPage > 0 ? (getPageUrl() + "?page=" + previousPage) : null;
    }

    public long getEndRange() {
        return Math.min(getItemCount(), getStartRange() + getItemsPerPage() - 1);
    }

    public long getStartRange() {
        return getIndex() + 1L;
    }

    public abstract List<V> getPageItems();
}

File: src/main/kotlin/javabot/web/views/PagedView.kt
package javabot.web.views

import javabot.dao.AdminDao
import javabot.dao.ApiDao
import javabot.dao.ChannelDao
import javabot.dao.FactoidDao
import javax.servlet.http.HttpServletRequest

abstract class PagedView<V>(
        adminDao: AdminDao,
        channelDao: ChannelDao,
        factoidDao: FactoidDao,
        apiDao: ApiDao,
        request: HttpServletRequest, private var page: Int) :
        MainView(adminDao, channelDao, factoidDao, apiDao, request) {
    private val itemsPerPage = ITEMS_PER_PAGE
    var itemCount: Long = -1
        get() {
            if (field == -1L) {
                field = countItems()
            }
            return field
        }


    override fun getChildView(): String {
        return "paged.ftl"
    }

    abstract fun getPagedView(): String

    abstract fun countItems(): Long

    fun getPage(): Int {
        if (page < 1) {
            page = 1
        } else if (page > getPageCount()) {
            page = getPageCount()
        }
        return page
    }

    fun getIndex(): Int {
        val index = (getPage() - 1) * getItemsPerPage()
        if (itemCount == 0L) {
            return -1
        } else if (index > itemCount) {
            return (getPageCount() - 1) * getItemsPerPage()
        } else {
            return index
        }
    }

    fun getPageCount(): Int {
        return Math.ceil(1.0 * itemCount / getItemsPerPage()).toInt()
    }

    fun getItemsPerPage(): Int {
        return itemsPerPage
    }

    open fun getNextPage(): String? {
        val nextPage = getPage() + 1
        return if (nextPage <= getPageCount()) getPageUrl() + "?page=" + nextPage else null
    }

    protected abstract fun getPageUrl(): String

    open fun getPreviousPage(): String? {
        val previousPage = if (getPage() == 1) -1 else getPage() - 1
        return if (previousPage > 0) (getPageUrl() + "?page=" + previousPage) else null
    }

    fun getEndRange(): Long {
        return Math.min(itemCount, getStartRange() + getItemsPerPage() - 1)
    }

    fun getStartRange(): Long {
        return getIndex() + 1L
    }

    abstract fun getPageItems(): List<V>

    companion object {
        val ITEMS_PER_PAGE: Int = 50
    }
}
